build-tools/xaprepare/xaprepare/Steps/Step_InstallDotNetPreview.cs

```diff
@@ -65,68 +62,188 @@ protected override async Task<bool> Execute (Context context)
 			return true;
 		}
 
-		async Task<bool> InstallDotNetAsync (Context context, string dotnetPath, string version, bool runtimeOnly = false)
+		async Task<bool> DownloadDotNetInstallScript (Context context, string dotnetScriptPath, Uri dotnetScriptUrl)
 		{
-			if (Directory.Exists (Path.Combine (dotnetPath, "sdk", version)) && !runtimeOnly) {
-				Log.Status ($"dotnet SDK version ");
-				Log.Status (version, ConsoleColor.Yellow);
-				Log.StatusLine (" already installed in: ", Path.Combine (dotnetPath, "sdk", version), tailColor: ConsoleColor.Cyan);
-				return true;
+			string tempDotnetScriptPath = dotnetScriptPath + "-tmp";
+			Utilities.DeleteFile (tempDotnetScriptPath);
+
+			Log.StatusLine ("Downloading dotnet-install...");
+
+			(bool success, ulong size, HttpStatusCode status) = await Utilities.GetDownloadSizeWithStatus (dotnetScriptUrl);
+			if (!success) {
+				string message;
+				if (status == HttpStatusCode.NotFound) {
+					message = "dotnet-install URL not found";
+				} else {
+					message = $"Failed to obtain dotnet-install size. HTTP status code: {status} ({(int)status})";
+				}
+
+				if (!ReportAndCheckCached (message)) {
+					return false;
+				}
 			}
 
-			if (Directory.Exists (Path.Combine (dotnetPath, "shared", "Microsoft.NETCore.App", version)) && runtimeOnly) {
-				Log.Status ($"dotnet runtime version ");
-				Log.Status (version, ConsoleColor.Yellow);
-				Log.StatusLine (" already installed in: ", Path.Combine (dotnetPath, "shared", "Microsoft.NETCore.App", version), tailColor: ConsoleColor.Cyan);
-				return true;
+			DownloadStatus downloadStatus = Utilities.SetupDownloadStatus (context, size, context.InteractiveSession);
+			Log.StatusLine ($"  {context.Characters.Link} {dotnetScriptUrl}", ConsoleColor.White);
+			await Download (context, dotnetScriptUrl, tempDotnetScriptPath, "dotnet-install", Path.GetFileName (dotnetScriptUrl.LocalPath), downloadStatus);
+
+			if (!File.Exists (tempDotnetScriptPath)) {
+				return ReportAndCheckCached ($"Download of dotnet-install from {dotnetScriptUrl} failed");
 			}
 
-			Uri dotnetScriptUrl = Configurables.Urls.DotNetInstallScript;
-			string dotnetScriptPath = Path.Combine (dotnetPath, Path.GetFileName (dotnetScriptUrl.LocalPath));
-			if (File.Exists (dotnetScriptPath))
-				Utilities.DeleteFile (dotnetScriptPath);
+			Utilities.CopyFile (tempDotnetScriptPath, dotnetScriptPath);
+			Utilities.DeleteFile (tempDotnetScriptPath);
+			return true;
 
-			Log.StatusLine ("Downloading dotnet-install...");
+			bool ReportAndCheckCached (string message)
+			{
+				if (File.Exists (dotnetScriptPath)) {
```

Comment:
It is downloaded unconditionally each time and refreshed in `~/android-archives/`, so the only time we'd use a "stale" copy is if there's no network access.

Reply:
I think we would need to return the value of `ReportAndCheckCached (message)` above on line 81 for this to work? The `await Utilities.GetDownloadSizeWithStatus (dotnetScriptUrl)` call on line 72 seems like the first thing that would fail if there is no network access, and it looks like that will return false all the way up.

That being said, does the dotnet install script even work to get installer URLs when offline?

Reply:
The script works offline if you pass a version to it other than `latest`, otherwise it will attempt to download https://dotnetcli.azureedge.net/dotnet/Sdk/LTS/latest.version and fail.

Regarding the `ReportAndCheckCached` call - it's possible that the size check will fail but network is there (it can e.g. return error 500) and the size check shouldn't be fatal in this case.

Reply:
Ok great, moved this comment to the appropriate place (line 82 above) and will resolve this one.